### PR TITLE
fix: escape backslashes in originalGit path

### PR DESCRIPTION
**Pull request by Tapppi**
Fix windows paths (and potentially others) by escaping
backslashes that are inserted into the mock-git js.

### CODE REVIEW

**Comment by stevemao** [comment]
Sorry but I'm not sure what this is trying to fix?

**Comment by Tapppi** [comment]
On windows `var originalGit = which('git');` comes out as `C:\xxxx\xxx\git\bin\git.exe`, and is then inserted into the JS string (`'childProcess.spawn("' + originalGit + '", argv, {'`) that is used as the mock. The backslashes end up as escape characters instead of actual backslashes in the spawn command. This results in spawn trying to spawn `C:xxxxxxgitbingit.exe`, which is obviously not found. The regex `\\` (double escaped in new RegExp because the string `\\\\` ends up as `\\` regexp) replaces a single backslash with double backslash, fixing this problem.

**Comment by stevemao** [comment]
But this would break paths like `/c/dodge-\\path` in other file systems?

**Comment by stevemao** [comment]
Also, I don't think a simple regex would be able to fix all file systems.

**Comment by Tapppi** [comment]
The problem isn't the file system. The problem is the escaping of the path. If I have the time today I might test this on *nix, but this should be a problem for paths like that as well. Unless I'm horribly mistaken, consider your example:
1. `/c/random-\path/git/bin/git` is the path of someone's git
2. They use `mock-git` which then gets that path from `which`.
3. The path is inserted into the mock JS as ...`childProcess.spawn("/c/random-\path/git/bin/git"`...
4. Due to standard string escaping, the spawn command will be seeing `\path` as an escape character `\`and `path`
5. This results in spawn using the path `/c/random-path/git/bin/git`, without the backslash, which is again incorrect.

In other words this really has nothing to with windows, but with an incorrectly escaped string instead.

**Comment by Tapppi** [comment]
@stevemao Added appveyor CI conf to this. It fails because https://github.com/stevemao/mock-bin needs to be released with the windows fixes. Do you still have questions regarding the logic in this PR?

**Comment by stevemao** [comment]
Do you think this is a bug in `shelljs`? Can you try `which` package and see if the problem persists?

**Comment by stevemao** [comment]
Look like it's a bug in `shelljs` where path is not correctly escaped on Windows.

**Comment by stevemao** [comment]
I manually merged 68135cccea5e292924f561db9f76b1222e06f227. Can you rebase?

**Comment by Tapppi** [comment]
I don't think that this falls into the scope of `shelljs` since it returns the backslash. The problem arises once it is inserted into the javascript string without escapes. For `shelljs` to return the string already escaped would be kind of weird for e.g. printing the result of which, no?

**Comment by Tapppi** [comment]
@stevemao Appveyor now runs correctly. This should pass on `mock-bin` release with the windows fixes.

**Comment by stevemao** [comment]
@Tapppi I published `mock-bin`, good job!
But the CI still fails here. But it doesn't look like this is the cause. Have a look at the latest builds. EG: https://ci.appveyor.com/project/stevemao/mock-git/build/11/job/ag5991dtfqimkrts

EDIT: the tests on this PR passed.

**Comment by Tapppi** [comment]
@stevemao Actually, this is the cause for that fail. The tests just don't do a very good job of pointing that out. They only test that the output of the original git doesn't match, **not that the original git is successfully run (stderr empty)**. That is also the reason this branch has the additional test asserts added. Checkout out this test I put together that has the original code, with additional asserts testing against `stderr` being empty, and with `shelljs` logging and a `console.log` for the `originalGit` from `which`:
https://ci.appveyor.com/project/Tapppi/mock-git/build/job/e619baxij8ihy3v8

The `shelljs` error output has the incorrect paths due to no escaping logged, and the assertions against `stderr` fail correctly.

This is also the reason this PR passed the tests now.

**Comment by stevemao** [comment]
Sorry, your right from the very beginning. Forgive my stupidity 😄

**Comment by Tapppi** [comment]
No problem, happy to contribute :relaxed:
